CLI done

JBoudMS opened the pull request:
- Has 3 commands - query, fetch-latest-batches, stream-latest-batches
- Stores/loads configuration from disk
- Interactively queries arguments if not provided and no default configuration set
- Takes arguments from the CLI and those have priority
- Encrypts credentials to disk
- Accepts client credentials, API Key, and JWTs
- Shows progress bar on fetches if set, disables console output unless specially set

metdockal commented on src/cli/auth-config.ts:
```diff
@@ -0,0 +1,104 @@
+	return Buffer.alloc(requiredKeyLength, key);
+}
+
+function encrypt(decryptedCredentials: string): { salt: string, iv: string, encryptedCredentials: string } {
+	const iv = randomBytes(16);
+	const salt = randomBytes(128).toString(BUFFER_ENCODING);
+	const key = getKeyFromPassword(salt, 32);
+	const cipher = createCipheriv(ENCRYPTION_ALGORITHM, Buffer.from(key), iv);
+	const encryptedCredentials = Buffer.concat([cipher.update(decryptedCredentials), cipher.final()]);
+	return {
+		salt,
+		iv: iv.toString(BUFFER_ENCODING),
+		encryptedCredentials: encryptedCredentials.toString(BUFFER_ENCODING)
+	};
+}
+
+function decrypt(encrypedCredentials: string, salt: string, iv: string) {
+	const key = getKeyFromPassword(salt, 32)
+	const decipher = createDecipheriv(ENCRYPTION_ALGORITHM, key, Buffer.from(iv, BUFFER_ENCODING));
+	let decrypted = decipher.update(Buffer.from(encrypedCredentials, BUFFER_ENCODING));
+	decrypted = Buffer.concat([decrypted, decipher.final()]);
+	return decrypted.toString();
+}
```
Why do you encrypt and decrypt credentials?

JBoudMS replied:
Just a layer of security. It doesn't add much, but means that in order for these credentials to leak they need to grab the passphrase out of our source control and decrypt them.

metdockal commented on src/cli/file-utils.ts:
```diff
@@ -0,0 +1,45 @@
+	}
+}
+
+async function writeConfig<T>(configType: ConfigType, configuration: T) {
```
in this case `configuration: any` would be 100% fine

metdockal commented on src/cli/auth-config.ts:
```diff
@@ -0,0 +1,104 @@
+
+const storeAuthenticationDetails = async (credentialDetails: CredentialDetails) => {
+	const authenticationConfig = transformCredentialDetailsIntoAuthenticationConfig(credentialDetails);
+	await writeConfig<AuthenticationConfig>(ConfigType.Credentials, authenticationConfig);
```
Same what I have raised last time :) This library is limited to NodeJS or CLI only

JBoudMS replied:
Yeah, suggestions on how to alleviate this would be good.

I've been thinking about splitting them into two. The executable wrapper doesn't need to live in the same package, but this just worsens what Troy higlighted: https://metocean.slack.com/archives/C014LE83VC5/p1602731385281800.

Thoughts?

JBoudMS replied:
I shouldn't think there should be too many extra merge conflicts, unless two developers are touching the same logical area at the same time. In which case you need to take extra precautions but we are already doing that.

I'm thinking we'll resolve it and resolve splitting this library into two once we've resolved it internally. We don't have clients trying to use the JS implementation yet, so there's no real downside yet.

metdockal commented on README.md:
```diff
@@ -23,10 +23,13 @@ We have not built any functionality to deal with this complexity below, instead
 # Using this library
 
-We provide a few ways to use this library depending on your use case. The up-to-date running versions of the code below lives in bin/examples.ts. We have extracted some common utility definitions such as the directory to store things and credentials. These are defined in the `bin/examples.ts` but are unnecessary noise below so have been removed.
+We provide a few ways to use this library depending on your use case. At it's simplest, this library provides an a CLI. If you run `npm install -g @metservice/global-lightning-client`, you can then run `npm metraweather-global-lightning`, which will walk you through configuring and using it. It will ask you to set some defaults, which can be overridden at any time.
```
you can also use npx (will always use the latest version) such as:
`npx -p @metservice/global-lightning-client metraweather-global-lightning`

metdockal reviewed (commented):
![image](https://user-images.githubusercontent.com/49377387/96519077-b424c580-12c8-11eb-953a-6e80b3098b51.png)

I cannot see anything in the console output

JBoudMS commented:
> ![image](https://user-images.githubusercontent.com/49377387/96519077-b424c580-12c8-11eb-953a-6e80b3098b51.png)
> 
> I cannot see anything in the console output

That's because it's fetching data a few days of data in 15 minute chunks, getting at most 10 strikes in each query. That's going to take a long time. It takes ~4 minutes to fetch the data from 2020-10-13T00:00:00.000Z to 2020-10-13T02:30:00.000Z with that limit. It will eventually complete, but I've had it running for 20 minutes and it's still going.

I've updated it to add a warning that reducing the values in both the limit and parallelQueries will reduce performance.

The current logic is to not show progress if you choose STDOUT or are not in interactive mode. That was because I mistakenly thought it would mess up the output if you piped it, but I was wrong. I've fixed that now so we always show the progress (it ends up in STDERR if you pipe STDOUT).

metdockal commented:
Ah.. true that wasn't clear to me (I didn't think about it much) :D
